js-api-spec/setup.ts

```diff
@@ -0,0 +1,387 @@
+      };
+    } else {
+      return {
+        message: () => 'expected callback not to throw a LegacyException',
+        pass: true,
+      };
+    }
+  }
+
+  return {
+    message: () => `expected ${received} to throw`,
+    pass: false,
+  };
+};
+
+const toEqualWithHash = (
+  received: unknown,
+  actual: immutable.ValueObject
+): SyncExpectationResult => {
+  if (typeof received !== 'object' || received === null) {
+    return {
+      message: () => `expected ${util.inspect(received)} to be an object`,
+      pass: false,
+    };
+  } else if (
+    !('equals' in received) ||
+    typeof (received as {equals: unknown}).equals !== 'function'
+  ) {
+    return {
+      message: () =>
+        `expected ${util.inspect(received)} to have an "equals" method`,
+      pass: false,
+    };
+  } else if (
+    !('hashCode' in received) ||
+    typeof (received as {hashCode: unknown}).hashCode !== 'function'
+  ) {
+    return {
+      message: () =>
+        `expected ${util.inspect(received)} to have a "hashCode" method`,
+      pass: false,
+    };
+  }
+
+  const receivedValue = received as immutable.ValueObject;
+  if (!receivedValue.equals(actual)) {
+    return {
+      message: () => `expected ${received} to be .equal() to ${actual}`,
+      pass: false,
+    };
+  } else if (receivedValue.hashCode() !== actual.hashCode()) {
+    return {
+      message: () =>
+        `expected ${received} to have the same .hashCode() as ${actual}`,
+      pass: false,
+    };
+  } else {
+    return {
+      message: () => `expected ${received} not to be .equal() to ${actual}`,
+      pass: true,
+    };
+  }
+};
+
+if (expect.extend) {
+  // Add custom matchers to Jest
+  expect.extend({
+    toThrowSassException,
+    toThrowLegacyException,
+    toEqualWithHash,
+  });
+} else if (jasmine?.addMatchers) {
+  // Add custom matchers to Jasmine
+  beforeAll(() => {
+    jasmine.addMatchers({
+      toThrowSassException: () => ({
+        compare: toThrowSassException,
+      }),
+      toThrowLegacyException: () => ({
+        compare: toThrowLegacyException,
+      }),
+      toEqualWithHash: () => ({
+        compare: toEqualWithHash,
+      }),
+      toHaveProperty: () => ({
+        compare: (
+          received: unknown,
+          expectedPath: string
+        ): SyncExpectationResult => {
+          if (typeof received !== 'object') {
+            throw new Error('Received value must be an object');
+          }
+          if (typeof expectedPath !== 'string') {
+            throw new Error('Expected property value must be a string');
+          }
+
+          return {
+            message: () =>
+              `expected ${received} to have property ${expectedPath}`,
+            pass: Object.prototype.hasOwnProperty.call(received, expectedPath),
+          };
+        },
+      }),
+    });
+    jasmine.addAsyncMatchers({
+      toThrowSassException: () => ({
+        compare: toThrowSassException,
+      }),
+    });
+  });
+}
```

Comment:
I think one of the main downsides of this current tooling is that while Jest and Jasmine have an almost identical API, some of the matchers are not 100% the same. Similar to the [default](https://jestjs.io/docs/expect) and [extended](https://github.com/jest-community/jest-extended) Jest matchers, this uses the [default](https://jasmine.github.io/api/4.6/matchers.html) and [extended](https://github.com/JamieMason/Jasmine-Matchers) Jasmine matchers -- but there are still occasional divergences.

Reply:
Would it be possible to just use Jasmine in _all_ the tests?

Reply:
@nex3 In https://github.com/sass/sass-spec/pull/1902/files/7219fb7ce424a9179c46c17d744f00859fa92b6d..ba3a0e2c6003367391d8dc545986e9e251f0e5a9 I converted all of the JS API tests to use Jasmine.

What do you think? (I could convert the `test/` directory as well, but since there's no overlap between `test/` and `js-api-spec/` I figured it didn't hurt to leave those running via Jest.)